optimade/models/jsonapi.py

```diff
@@ -1,81 +1,98 @@
-"""
-This module should reproduce https://jsonapi.org/schema
-"""
-from typing import Optional, Set, Union, Dict, Any
-from pydantic import BaseModel, UrlStr, constr, Schema
-
-
-class Meta(Dict[str, Any]):
+"""This module should reproduce JSON API v1.0 https://jsonapi.org/format/1.0/"""
+from typing import Optional, Set, Union
+from pydantic import BaseModel, UrlStr, Schema, validator
+
+
+__all__ = (
+    "Meta",
+    "Link",
+    "JsonApi",
+    "ToplevelLinks",
+    "ErrorLinks",
+    "ErrorSource",
+    "BaseResource",
+    "RelationshipLinks",
+    "Relationship",
+    "Relationships",
+    "ResourceLinks",
+    "Attributes",
+    "Resource",
+    "Response",
+)
+
+
+class Meta(BaseModel):
     """Non-standard meta-information that can not be represented as an attribute or relationship."""
 
+    class Config:
+        extra = "allow"
+
 
 class Link(BaseModel):
     """A link **MUST** be represented as either: a string containing the link's URL or a link object."""
 
     href: UrlStr = Schema(..., description="a string containing the link’s URL.")
-    meta: Optional[dict] = Schema(
+    meta: Optional[Meta] = Schema(
         ...,
         description="a meta object containing non-standard meta-information about the link.",
     )
 
 
-class Links(BaseModel):
-    """A Links object is a set of keys with a Link value"""
-
-    next: Optional[Union[UrlStr, Link]] = Schema(
-        ..., description="A Link to the next object"
-    )
-    self: Optional[Union[UrlStr, Link]] = Schema(..., description="A link to itself")
-    related: Optional[Union[UrlStr, Link]] = Schema(
-        ..., description="A related resource link"
-    )
-    about: Optional[Union[UrlStr, Link]] = Schema(
-        ...,
-        description="a link that leads to further details about this particular occurrence of the problem.",
-    )
-
-
-class JsonAPI(BaseModel):
+class JsonApi(BaseModel):
     """An object describing the server's implementation"""
 
     version: str = Schema(..., description="Version of the json API used")
-    meta: Optional[dict] = Schema(..., description="Non-standard meta information")
+    meta: Optional[Meta] = Schema(..., description="Non-standard meta information")
 
 
-class Pagination(BaseModel):
-    """
-    A set of urls to different pages:
-    """
+class ToplevelLinks(BaseModel):
```

Comment:
Even though this is valid as part of the JSON API v1.0 models, it can also be moved to `toplevel` models in order to avoid implementors importing any model from `jsonapi`?

Reply:
I'm fine with either solution: moving this to `toplevel` or importing this from `jsonapi` as the one exception to the rule.